### PR TITLE
Added stubs for discovery7.0 feature from UI/CLI/API

### PR DESCRIPTION
**Pull request by sghai**
- Added test stubs based on changes/bug fixes in discovery7.0
- Added all tests for discovery-rules from CLI

### CODE REVIEW

**Comment by sghai** [comment]
referencing issue #3806

File: tests/foreman/ui/test_discoveryrule.py [hunk 1]
```diff
@@ -634,3 +640,109 @@ def test_negative_update_priority(self):
...
+    @run_only_on('sat')
+    @stubbed()
+    @tier2
+    def test_positive_view_existing_rule_with_non_admin_user(self):
+        """Existing rule should be viewed to non-admin user by associating
+        discovery_reader role.
+
+        @id: 0f5b0221-43be-47bc-8619-749824c4e54f
+
+        @Step:
```
Comment (rochacbruno) [inline tests/foreman/ui/test_discoveryrule.py]: typo, `@steps`

**Comment by rochacbruno** [comment]
ACK (pending a comment and travis build)

**Comment by sghai** [comment]
Ping ? Please review the PR :)

File: tests/foreman/api/test_discoveryrule.py [hunk 2]
```diff
@@ -77,6 +77,19 @@ def test_positive_create_with_name(self):
                     discovery_rule.search_,
                 )
 
+    @run_only_on('sat')
+    @stubbed()
+    @tier1
+    def test_positive_create_by_associating_org_loc(self):
```
Comment (abalakh) [inline tests/foreman/api/test_discoveryrule.py]: `create_by_associating_org_loc` or maybe `create_with_org_loc`?

File: tests/foreman/api/test_discoveryrule.py [hunk 3]
```diff
@@ -150,6 +163,19 @@ def test_positive_update_name(self):
                 discovery_rule = discovery_rule.update(['name'])
                 self.assertEqual(discovery_rule.name, name)
 
+    @run_only_on('sat')
+    @stubbed()
+    @tier1
+    def test_positive_update_by_associating_org_loc(self):
```
Comment (abalakh) [inline tests/foreman/api/test_discoveryrule.py]: Same here. How about `test_positive_add_org_loc`?

File: tests/foreman/cli/test_discoveryrule.py [hunk 4]
```diff
@@ -0,0 +1,422 @@
...
+
+        @Assert: Rule was created and with given org & location names.
+
+        @caseautomation: notautomated
+        """
+
+    @run_only_on('sat')
+    @stubbed()
+    @tier1
+    def test_positive_create_by_associating_org_loc_id(self):
```
Comment (abalakh) [inline tests/foreman/cli/test_discoveryrule.py]: > maybe `create_with_org_loc`?

Comment (abalakh) [inline tests/foreman/cli/test_discoveryrule.py]: btw, as you mentioned `org_loc_id` in name, i assume there's another way, by specifying org and loc names? If so, please also add test stubs for that

File: tests/foreman/cli/test_discoveryrule.py [hunk 5]
```diff
@@ -0,0 +1,422 @@
...
+        @Assert: Rule should be successfully created and has expected hostname
+        value
+
+        @caseautomation: notautomated
+        """
+
+    @run_only_on('sat')
+    @stubbed()
+    @tier1
+    def test_positive_create_by_associating_org_loc_name(self):
```
Comment (abalakh) [inline tests/foreman/cli/test_discoveryrule.py]: > maybe `test_positive_create_with_org_loc_by_name`?

File: tests/foreman/cli/test_discoveryrule.py [hunk 6]
```diff
@@ -0,0 +1,422 @@
...
+        @id: 8b7411b8-bb4b-483e-837b-c468620ff99b
+
+        @Assert: Rule was created and with given org & location ids.
+
+        @caseautomation: notautomated
+        """
+
+    @run_only_on('sat')
+    @tier1
+    def test_positive_create_with_hosts_limit(self):
```
Comment (abalakh) [inline tests/foreman/cli/test_discoveryrule.py]: missing `@stubbed`

File: tests/foreman/cli/test_discoveryrule.py [hunk 7]
```diff
@@ -0,0 +1,422 @@
...
+
+        @Assert: Rule should be successfully created and has expected hosts
+        limit value
+
+        @caseautomation: notautomated
+        """
+
+    @run_only_on('sat')
+    @tier1
+    def test_positive_create_with_max_count(self):
```
Comment (abalakh) [inline tests/foreman/cli/test_discoveryrule.py]: > missing `@stubbed`

File: tests/foreman/cli/test_discoveryrule.py [hunk 8]
```diff
@@ -0,0 +1,422 @@
...
+
+        @Assert: Rule should be successfully created and has max_count set
+        as per given value
+
+        @caseautomation: notautomated
+        """
+
+    @run_only_on('sat')
+    @tier1
+    def test_positive_create_with_priority(self):
```
Comment (abalakh) [inline tests/foreman/cli/test_discoveryrule.py]: > missing `@stubbed`

File: tests/foreman/cli/test_discoveryrule.py [hunk 9]
```diff
@@ -0,0 +1,422 @@
...
+
+        @Assert: Rule should be successfully created and has expected priority
+        value
+
+        @caseautomation: notautomated
+        """
+
+    @run_only_on('sat')
+    @tier1
+    def test_positive_create_disabled_rule(self):
```
Comment (abalakh) [inline tests/foreman/cli/test_discoveryrule.py]: > missing `@stubbed`

File: tests/foreman/cli/test_discoveryrule.py [hunk 10]
```diff
@@ -0,0 +1,422 @@
...
+        @id: 8837a0c6-e19a-4c33-8b87-07b6f69dbb0f
+
+        @Assert: Disabled rule should be successfully created
+
+        @caseautomation: notautomated
+        """
+
+    @run_only_on('sat')
+    @tier1
+    def test_negative_create_with_invalid_name(self):
```
Comment (abalakh) [inline tests/foreman/cli/test_discoveryrule.py]: > missing `@stubbed`

File: tests/foreman/cli/test_discoveryrule.py [hunk 11]
```diff
@@ -0,0 +1,422 @@
...
+        @id: a0350dc9-8f5b-4673-be88-a5e35d1f8ca7
+
+        @Assert: Error should be raised and rule should not be created
+
+        @caseautomation: notautomated
+        """
+
+    @run_only_on('sat')
+    @tier1
+    def test_negative_create_with_invalid_hostname(self):
```
Comment (abalakh) [inline tests/foreman/cli/test_discoveryrule.py]: > missing `@stubbed`

File: tests/foreman/cli/test_discoveryrule.py [hunk 12]
```diff
@@ -0,0 +1,422 @@
...
+        @id: 0ae51085-30d0-44f9-9e49-abe928a8a4b7
+
+        @Assert: Error should be raised and rule should not be created
+
+        @caseautomation: notautomated
+        """
+
+    @run_only_on('sat')
+    @tier1
+    def test_negative_create_with_too_long_limit(self):
```
Comment (abalakh) [inline tests/foreman/cli/test_discoveryrule.py]: > missing `@stubbed`

File: tests/foreman/cli/test_discoveryrule.py [hunk 13]
```diff
@@ -0,0 +1,422 @@
...
+
+        @Assert: Validation error should be raised and rule should not be
+        created
+
+        @caseautomation: notautomated
+        """
+
+    @run_only_on('sat')
+    @tier1
+    def test_negative_create_with_same_name(self):
```
Comment (abalakh) [inline tests/foreman/cli/test_discoveryrule.py]: > missing `@stubbed`

File: tests/foreman/cli/test_discoveryrule.py [hunk 14]
```diff
@@ -0,0 +1,422 @@
...
+        @id: 5a914e76-de01-406d-9860-0e4e1521b074
+
+        @Assert: Error should be raised and rule should not be created
+
+        @caseautomation: notautomated
+        """
+
+    @run_only_on('sat')
+    @tier1
+    def test_positive_delete(self):
```
Comment (abalakh) [inline tests/foreman/cli/test_discoveryrule.py]: > missing `@stubbed`

File: tests/foreman/cli/test_discoveryrule.py [hunk 15]
```diff
@@ -0,0 +1,422 @@
...
+        @id: c9b88a94-13c4-496f-a5c1-c088187250dc
+
+        @Assert: Rule should be successfully deleted
+
+        @caseautomation: notautomated
+        """
+
+    @run_only_on('sat')
+    @tier1
+    def test_positive_update_name(self):
```
Comment (abalakh) [inline tests/foreman/cli/test_discoveryrule.py]: > missing `@stubbed`

File: tests/foreman/cli/test_discoveryrule.py [hunk 16]
```diff
@@ -0,0 +1,422 @@
...
+
+        @Assert: Rule name is updated
+
+        @caseautomation: notautomated
+        """
+
+    @run_only_on('sat')
+    @stubbed()
+    @tier1
+    def test_positive_update_by_associating_org_loc(self):
```
Comment (abalakh) [inline tests/foreman/cli/test_discoveryrule.py]: `test_positive_add_org_loc`?

Comment (abalakh) [inline tests/foreman/cli/test_discoveryrule.py]: Btw, apart from adding org and loc, don't we need a test for updating existing org and loc? Just a suggestion

File: tests/foreman/ui/test_discoveredhost.py [hunk 17]
```diff
@@ -950,3 +950,224 @@ def test_negative_pxe_less_discovery_without_dhcp(self):
...
+
+        @caseautomation: notautomated
+
+        @CaseLevel: System
+        """
+
+    @run_only_on('sat')
+    @stubbed()
+    @tier3
+    def test_positive_keep_facts_set_by_user(self):
```
Comment (abalakh) [inline tests/foreman/ui/test_discoveredhost.py]: Basically main action of the test is missing in the test name. How about test_positive_*provision*_with_facts_set_by_user?

File: tests/foreman/ui/test_discoveredhost.py [hunk 18]
```diff
@@ -950,3 +950,224 @@ def test_negative_pxe_less_discovery_without_dhcp(self):
...
+
+        @caseautomation: notautomated
+
+        @CaseLevel: System
+        """
+
+    @run_only_on('sat')
+    @stubbed()
+    @tier3
+    def test_positive_clear_facts_set_by_user(self):
```
Comment (abalakh) [inline tests/foreman/ui/test_discoveredhost.py]: `test_positive_provision_clear_facts_set_by_user` or `test_positive_provision_without_facts_set_by_user`

File: tests/foreman/api/test_discoveryrule.py (hunk 3 above)
Comment (sghai) [inline tests/foreman/api/test_discoveryrule.py]: test is to update org and location. so will update with: `test_positive_update_with_org_loc`

File: tests/foreman/cli/test_discoveryrule.py (hunk 16 above)
Comment (sghai) [inline tests/foreman/cli/test_discoveryrule.py]: Yeah.. the test pointed above is for updating org_location only. So will update with name:
`test_positive_update_org_loc`

File: tests/foreman/cli/test_discoveryrule.py [hunk 19]
```diff
@@ -0,0 +1,422 @@
...
+
+        @id: 86943095-acc5-40ff-8e3c-88c76b36333d
+
+        @Assert: Rule search field is updated
+
+        @caseautomation: notautomated
+        """
+
+    @run_only_on('sat')
+    @tier1
```
Comment (sghai) [inline tests/foreman/cli/test_discoveryrule.py]: should be tier1 only as we are just updating existing search query.. no new association here

**Comment by abalakh** [comment]
ACK

**Pull request merged**